python/src/main/python/drivers/graphicsfuzz_piglit_converter.py

```diff
@@ -0,0 +1,225 @@
+    :return: the shader_test vertex header string.
+    """
+    return VERTEX_HEADER + '\n'
+
+
+def make_fragment_shader_header(fragment_shader: str) -> str:
+    """
+    Creates the piglit [fragment shader] header.
+    :param fragment_shader: the fragment shader code.
+    :return: the shader_test fragment header string.
+    """
+    frag_header = FRAGMENT_HEADER + '\n'
+    frag_header += fragment_shader
+    return frag_header
+
+
+def make_test_header(shader_job_json_parsed: dict) -> str:
+    """
+    Creates the [test] header. Loads uniforms based on the uniforms found in the shader.
+    Note that piglit really doesn't like it if you try to load a uniform that isn't actually
+    used in the shader - because of this, we actually have to parse the shader itself and
+    find out which uniforms are being used instead of blindly pasting from the shader job JSON file.
+    :param shader_job_json_parsed: the parsed JSON properties.
+    :return: the shader_test test header string.
+    """
+    test_header = TEST_HEADER + '\n'
+    for key, value in shader_job_json_parsed.items():
+        test_header += UNIFORM_DEC + ' '
+        uniform_type = value['func']
+        if uniform_type not in UNIFORM_TYPES.keys():
+            raise AssertionError('Unknown uniform type: ' + uniform_type)
+        test_header += UNIFORM_TYPES[uniform_type] + ' ' + str(key) + ' '
+        test_header += str(value['args'][0:]).strip('[]').replace(',', '') + '\n'
+    test_header += DRAW_COMMAND
+    return test_header
+
+
+def is_uniform(line: str) -> bool:
+    """
+    Helper function to see if a given string is a GLSL uniform declaration.
+    :param line: the line to check
+    :return: True if the line is a uniform declaration, False otherwise.
+    """
+    return 'uniform' in line
+
+
+def is_version_header(line: str) -> bool:
+    """
+    Helper function to see if a given string is a GLSL preprocessor version string.
+    :param line: the line of code to check
+    :return: True if the string is a GLSL preprocessor version string, False otherwise.
+    """
+    return SHADER_VERSION_FLAG in line
+
+
+def get_json_properties(shader_job: str) -> List:
+    """
+    Helper function to parse a shader job JSON file into a list of properties.
+    Throws IOError if the file can't be parsed.
+    :param shader_job: the path to the shader job file.
+    :return: a list of JSON properties.
+    """
+    with open(shader_job, 'r', encoding='utf-8', errors='ignore') as job:
+        json_parsed = json.load(job)
+    if not json_parsed:
+        raise IOError('Malformed shader job file.')
+    return json_parsed
+
+
+def get_shader_from_job(shader_job: str) -> str:
+    """
+    Helper function to get the filename of a shader file from its corresponding shader job.
+    :param shader_job: the path of the shader job file
+    :return: the path of the fragment shader file.
+    """
+    return shader_job.split('.')[0] + '.frag'
```

Comment:
This could fail if the filename is e.g. 'my.shader.job.json'. I think we use a function called `remove_end` for this; feel free to import it from runspv.

Reply:
By the way, the Python (3) approach is probably to use pathlib.Path (and then the `.stem` attribute to get the path without the extension) . But Python has a weird mix of usually expecting paths as strings, and sometimes allowing the use of Path. Since you did not use Path from the beginning, it is probably not worth introducing it now.

Reply:
For some reason I thought when writing this that periods were only allowed as the file ending suffix in paths, so that's why I went with that. Thanks for catching this.